backend/inspirehep/tools/views.py

```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2020 CERN.
+#
+# inspirehep is free software; you can redistribute it and/or modify it under
+# the terms of the MIT License; see LICENSE file for more details.
+
+from io import BytesIO, TextIOWrapper
+
+from flask import Blueprint, abort, current_app, jsonify, request
+
+from inspirehep.files.api import current_s3_instance
+from inspirehep.records.utils import hash_data
+from inspirehep.tools.utils import get_filename, get_mimetype, get_references
+
+blueprint = Blueprint("inspirehep_tools", __name__, url_prefix="/tools")
+
+ALLOWED_EXTENSIONS = {"tex"}
+
+
+def allowed_file(filename):
+    return "." in filename and filename.rsplit(".", 1)[1].lower() in ALLOWED_EXTENSIONS
+
+
+@blueprint.route("/bibgen", methods=["POST"])
+def generate_bibliography():
+
+    requested_format = request.args.get("format")
+
+    bytes_file = request.files["file"]
+
+    if not allowed_file(bytes_file.filename):
+        abort(400)
+
+    text_file = TextIOWrapper(bytes_file)
+
+    references, errors = get_references(text_file, requested_format)
+
+    file_data = "\n\n".join(references).encode("utf-8")
+    key = hash_data(file_data)
+    file_data = BytesIO(file_data)
+    filename = get_filename(bytes_file.filename, requested_format)
+    mime_type = get_mimetype(requested_format)
+    bucket = current_app.config.get("S3_BIBGEN_BUCKET")
+
+    current_s3_instance.upload_file(
+        file_data, key, filename, mime_type, current_app.config["S3_FILE_ACL"], bucket
+    )
+    download_url = f"{current_app.config.get('S3_HOSTNAME')}/{bucket}/{key}"
+
+    for error in errors:
+        error = {
+            "message": f"Reference to {error['ref']} not found on line {error['line']}."
```

Comment:
Confusing error mesage, as sometimes there are too many found references. These two cases should be distinguished.

Reply:
The refinement didn't differentiate between both cases, but I can change this, I also think it'd be better and easy.